ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -68,43 +72,60 @@ bech32Polymod values = foldl' go 1 values .&. 0x3fffffff
 
-bech32CreateChecksum :: HRP -> [Word5] -> [Word5]
-bech32CreateChecksum hrp dat = [word5 (polymod .>>. i) | i <- [25,20..0]]
+bech32CreateChecksum :: Word -> HRP -> [Word5] -> [Word5]
+bech32CreateChecksum spec hrp dat = [word5 (polymod .>>. i) | i <- [25,20..0]]
   where
     values = bech32HRPExpand hrp ++ dat
-    polymod = bech32Polymod (values ++ map UnsafeWord5 [0, 0, 0, 0, 0, 0]) `xor` 1
+    polymod = bech32Polymod (values ++ map UnsafeWord5 [0, 0, 0, 0, 0, 0]) `xor` spec
+
+bech32VerifyChecksum :: HRP -> [Word5] -> Word
+bech32VerifyChecksum hrp dat = bech32Polymod (bech32HRPExpand hrp ++ dat)
 
-bech32VerifyChecksum :: HRP -> [Word5] -> Bool
-bech32VerifyChecksum hrp dat = bech32Polymod (bech32HRPExpand hrp ++ dat) == 1
+data EncodeError =
+      ResultStringLengthExceeded
+    | InvalidHumanReadablePart
+    deriving (Show, Eq)
 
-bech32Encode :: HRP -> [Word5] -> Maybe BS.ByteString
-bech32Encode hrp dat = do
-    guard $ checkHRP hrp
-    let dat' = dat ++ bech32CreateChecksum hrp dat
+bech32Encode :: Word -> HRP -> [Word5] -> Either EncodeError BS.ByteString
+bech32Encode spec hrp dat = do
+    verify InvalidHumanReadablePart $ validHRP hrp
+    let dat' = dat ++ bech32CreateChecksum spec hrp dat
         rest = map (charset Arr.!) dat'
         result = BSC.concat [BSC.map toLower hrp, BSC.pack "1", BSC.pack rest]
-    guard $ BS.length result <= 90
+    verify ResultStringLengthExceeded $ BS.length result <= 90
     return result
 
-checkHRP :: BS.ByteString -> Bool
-checkHRP hrp = not (BS.null hrp) && BS.all (\char -> char >= 33 && char <= 126) hrp
+validHRP :: BS.ByteString -> Bool
+validHRP hrp = not (BS.null hrp) && BS.all (\char -> char >= 33 && char <= 126) hrp
 
-bech32Decode :: BS.ByteString -> Maybe (HRP, [Word5])
+data DecodeError =
+    Bech32StringLengthExceeded
+  | CaseInconsistency
+  | TooShortDataPart
+  | InvalidHRP
+  | ChecksumVerificationFail
+  | InvalidCharsetMap
+  deriving (Show, Eq)
+
+bech32Decode :: BS.ByteString -> Either DecodeError (Word, HRP, [Word5])
 bech32Decode bech32 = do
-    guard $ BS.length bech32 <= 90
-    guard $ BSC.map toUpper bech32 == bech32 || BSC.map toLower bech32 == bech32
+    verify Bech32StringLengthExceeded $ BS.length bech32 <= 90
+    verify CaseInconsistency $ validCase bech32
     let (hrp, dat) = BSC.breakEnd (== '1') $ BSC.map toLower bech32
-    guard $ BS.length dat >= 6
-    hrp' <- BSC.stripSuffix (BSC.pack "1") hrp
-    guard $ checkHRP hrp'
-    dat' <- mapM charsetMap $ BSC.unpack dat
-    guard $ bech32VerifyChecksum hrp' dat'
-    return (hrp', take (BS.length dat - 6) dat')
+    verify TooShortDataPart $ BS.length dat >= 6
+    hrp' <- maybeToRight InvalidHRP $ BSC.stripSuffix (BSC.pack "1") hrp
+    verify InvalidHRP $ validHRP hrp'
+    dat' <- maybeToRight InvalidCharsetMap . mapM charsetMap $ BSC.unpack dat
+    let spec = bech32VerifyChecksum hrp' dat'
```

Comment:
μ-nit: Might just be me, but I feel like a "Verify" function should either pass or fail, rather than return the checksum constant for future comparison. Perhaps bech32Decode should just try both encoding constants and return the encoding type if one of them succeeds

Reply:
I've renamed `bech32VerifyCheckum` as `bech32Residue`.